core/types/src/attestation_key.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Attestation Key types
+
+use crate::{
+    attestation_key::QuoteSignature::{Linkable, UnLinkable},
+    new_type_accessors_impls,
+    report::{ExtendedProductId, FamilyId},
+    ConfigId, FfiError,
+};
+use mc_sgx_core_sys_types::{sgx_att_key_id_ext_t, sgx_ql_att_key_id_t, sgx_quote_sign_type_t};
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum MrSignerKeyHash {
+    Sha256([u8; 32]),
+    Sha384([u8; 48]),
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct AlgorithmId(u32);
+
+new_type_accessors_impls! {
+    AlgorithmId, u32;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Version(u16);
+
+new_type_accessors_impls! {
+    Version, u16;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Id(u16);
+
+new_type_accessors_impls! {
+    Id, u16;
+}
+
+/// The type of quote signature.  Only valid for EPID quotes.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+#[non_exhaustive]
+#[repr(u16)]
+enum QuoteSignature {
+    UnLinkable,
+    Linkable,
+}
+
+impl From<sgx_quote_sign_type_t> for QuoteSignature {
+    fn from(sign_type: sgx_quote_sign_type_t) -> QuoteSignature {
+        match sign_type {
+            // Per the header `sgx_quote.h` and
+            // https://download.01.org/intel-sgx/sgx-linux/2.17.1/docs/Intel_SGX_Developer_Reference_Linux_2.17.1_Open_Source.pdf
+            // the `sgx_att_key_id_ext_t::att_key_type` is only valid for EPID
+            // quotes it will be 0 otherwise, which also happens to map to
+            // SGX_UNLINKABLE_SIGNATURE.
+            sgx_quote_sign_type_t::SGX_UNLINKABLE_SIGNATURE => UnLinkable,
+            sgx_quote_sign_type_t::SGX_LINKABLE_SIGNATURE => Linkable,
+            v => panic!("Got unknown 'sgx_quote_sign_type_t': {:?}", v),
```

Comment:
I'm not sure if a `panic!` here is correct, since because of C/C++ weirdness it's quite possible for a variable of "type `sgx_quote_sign_type_t` to contain a value that's not one of the enumerated values. We might prefer to use `TryFrom` instead.

Reply:
converted to `TryFrom`